python/kotekan/config.py

```diff
@@ -0,0 +1,163 @@
+        If the file is .j2 and cannot be parsed by jinja2.
+    """
+    file_ext = os.path.splitext(file_name_full)[1]
+    directory, file_name = os.path.split(file_name_full)
+
+    if yaml is None:
+        raise ImportError("yaml is required for parsing configuration files.")
+
+    # Treat all files as pure YAML, unless it is a ".j2" file, then run jinja.
+    if file_ext != ".j2":
+        # Lint the YAML file, helpful for finding errors
+        try:
+            output = subprocess.Popen(
+                [
+                    "yamllint",
+                    "-d",
+                    "{extends: relaxed, \
+                                         rules: {line-length: {max: 100}, \
+                                                commas: disable, \
+                                                brackets: disable, \
+                                                trailing-spaces: {level: warning}}}",
+                    file_name_full,
+                ],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+            )
+            response, stderr = output.communicate()
+            if response != "":
+                sys.stderr.write("yamllint warnings/errors for: ")
+                sys.stderr.write(str(response))
+        # TODO: change to checking for OSError subtypes when Python 2 support is removed
+        except OSError as e:
+            if e.errno == errno.ENOENT:
+                sys.stderr.write("yamllint not installed, skipping pre-validation\n")
+            else:
+                sys.stderr.write("error with yamllint, skipping pre-validation\n")
+
+        try:
+            with open(file_name_full, "r") as stream:
+                config_yaml = yaml.safe_load(stream)
+        except IOError as err:
+            raise IOError(
+                "Error reading file " + file_name_full + ": " + str(err)
+            ) from err
+        except yaml.YAMLError as err:
+            raise yaml.YAMLError("Error parsing yaml: \n" + str(err) + "\n") from err
+
+        if dump:
+            sys.stderr.write(yaml.dump(config_yaml) + "\n")
+
+    else:
+        if jinja2 is None:
+            raise ImportError("jinja2 is required for processing '.j2' files")
+
+        # Load the template
+        env = jinja2.Environment(
+            loader=jinja2.FileSystemLoader(directory),
+            autoescape=jinja2.select_autoescape(),
+        )
+        template = env.get_template(file_name)
+
+        # Parse the optional variables (if any)
+        options_dict = {}
+        if jinja_options:
+            options_dict = json.loads(str(jinja_options))
+
+        # Convert to yaml
+        config_yaml_raw = template.render(options_dict)
+
+        # Dump the rendered yaml file if requested
+        if dump:
+            sys.stderr.write(config_yaml_raw + "\n")
+
+        # TODO Should we also lint the output of the template?
+        config_yaml = yaml.safe_load(config_yaml_raw)
```

Comment:
This should be in a `try:` block to catch `yaml.YAMLError` and output the parsing error as above and in the original script.

Reply:
Why do we want to parse the error and print to stdout instead of just allowing the YAMLError to be raised normally?

Reply:
Because when called within `kotekan.cpp` the error message doesn't make it to command line of the person running the code.  I think this was because kotekan assumes the config will come via STDOUT, and the python error message from the exception also went to STDOUT.   This resulted in us not seeing what line number/reason the error was generating the problem - just a JSON parse error from kotekan.  Given we've moved this to a library + script design, perhaps the correct way to deal with this is to allow it raise the exception here, but in the script part use a `try:` block to catch the exception and write out the err string details to STDERR and exit the script without writing anything to STDOUT.

Reply:
That makes sense. I just pushed some changes that add an option to the script mode of `config.py`, which will write out exceptions to stderr instead of stdout.